File: app/src/main/java/com/czbix/v2ex/model/Topic.java
package com.czbix.v2ex.model;

import android.os.Parcel;

import com.czbix.v2ex.common.exception.FatalException;
import com.czbix.v2ex.network.RequestHelper;
import com.google.common.base.Objects;
import com.google.common.base.Preconditions;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Topic extends Page implements ThankAble, IgnoreAble {
    private static final Pattern PATTERN = Pattern.compile("/t/(\\d+?)(?:\\W|$)");

    private final int mId;
    private final String mTitle;
    private final String mContent;
    private final int mReplies;
    private final Member mMember;
    private final Node mNode;
    private final String mReplyTime;
    private final boolean mHasInfo;
    private boolean mHasRead;

    Topic(String title, int id, String content, Member member, Node node, String replyTime, int replies) {
        Preconditions.checkArgument(id != 0);

        mId = id;

        mTitle = title;
        mContent = content;
        mMember = member;
        mNode = node;
        mReplies = replies;
        mReplyTime = replyTime;

        mHasInfo = member != null;
    }

    public boolean hasInfo() {
        return mHasInfo;
    }

    public int getId() {
        return mId;
    }

    public String getReplyTime() {
        return mReplyTime;
    }

    @Override
    public String getTitle() {
        return mTitle;
    }

    public Member getMember() {
        return mMember;
    }

    public Node getNode() {
        return mNode;
    }

    public int getReplyCount() {
        return mReplies;
    }

    public static String buildUrlFromId(int id) {
        return RequestHelper.BASE_URL + "/t/" + Integer.toString(id);
    }

    public static int getIdFromUrl(String url) {
        final Matcher matcher = PATTERN.matcher(url);
        if (!matcher.find()) {
            throw new FatalException("match id for topic failed: " + url);
        }
        final String idStr = matcher.group(1);
        return Integer.parseInt(idStr);
    }

    public String getContent() {
        return mContent;
    }

    @Override
    public String getUrl() {
        return buildUrlFromId(mId);
    }

    @Override
    public String getIgnoreUrl() {
        return String.format("%s/ignore/topic/%d", RequestHelper.BASE_URL, mId);
    }

    @Override
    public String getThankUrl() {
        return String.format("%s/thank/topic/%d", RequestHelper.BASE_URL, mId);
    }

    public boolean hasRead() {
        return mHasRead;
    }

    public void setHasRead() {
        mHasRead = true;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(this.mId);
        dest.writeByte(mHasInfo ? (byte) 1 : (byte) 0);
        dest.writeString(this.mTitle);
        dest.writeString(this.mContent);
        dest.writeInt(this.mReplies);
        dest.writeParcelable(this.mMember, 0);
        dest.writeParcelable(this.mNode, 0);
        dest.writeString(this.mReplyTime);
    }

    protected Topic(Parcel in) {
        this.mId = in.readInt();
        this.mHasInfo = in.readByte() != 0;
        this.mTitle = in.readString();
        this.mContent = in.readString();
        this.mReplies = in.readInt();
        this.mMember = in.readParcelable(Member.class.getClassLoader());
        this.mNode = in.readParcelable(Node.class.getClassLoader());
        this.mReplyTime = in.readString();
    }

    public static final Creator<Topic> CREATOR = new Creator<Topic>() {
        public Topic createFromParcel(Parcel source) {
            return new Topic(source);
        }

        public Topic[] newArray(int size) {
            return new Topic[size];
        }
    };

    public Builder toBuilder() {
        return new Builder()
                .setId(mId)
                .setTitle(mTitle)
                .setContent(mContent)
                .setMember(mMember)
                .setNode(mNode)
                .setReplyCount(mReplies)
                .setReplyTime(mReplyTime);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Topic)) return false;
        Topic topic = (Topic) o;
        return Objects.equal(mId, topic.mId) &&
                Objects.equal(mReplies, topic.mReplies) &&
                Objects.equal(mContent, topic.mContent) &&
                Objects.equal(mNode, topic.mNode) &&
                Objects.equal(mReplyTime, topic.mReplyTime);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(mId, mContent, mReplies, mNode, mReplyTime);
    }

    public static class Builder {
        private int mId;
        private String mTitle;
        private String mContent;
        private Member mMember;
        private Node mNode;
        private int mReplies;
        private String mReplyTime;

        public Builder setTitle(String title) {
            mTitle = title;
            return this;
        }

        public Builder setId(int id) {
            mId = id;
            return this;
        }

        public Builder setContent(String content) {
            mContent = content;
            return this;
        }

        public Builder setMember(Member member) {
            mMember = member;
            return this;
        }

        public Builder setNode(Node node) {
            mNode = node;
            return this;
        }

        public Builder setReplyCount(int replies) {
            mReplies = replies;
            return this;
        }

        public Builder setReplyTime(String replyTime) {
            mReplyTime = replyTime;
            return this;
        }

        public boolean hasInfo() {
            return mMember != null;
        }

        public Topic createTopic() {
            return new Topic(mTitle, mId, mContent, mMember, mNode, mReplyTime, mReplies);
        }
    }
}

File: app/src/main/kotlin/com/czbix/v2ex/model/Topic.kt
package com.czbix.v2ex.model

import com.czbix.v2ex.db.Member
import com.czbix.v2ex.network.RequestHelper
import kotlinx.android.parcel.IgnoredOnParcel
import kotlinx.android.parcel.Parcelize
import java.util.regex.Pattern
import kotlin.properties.Delegates

@Parcelize
data class Topic(
        val id: Int,
        @JvmField
        val title: String?,
        val content: List<ContentBlock>?,
        val replyCount: Int,
        val member: Member?,
        val node: Node?,
        val replyTime: String?,
        val isFavored: Boolean,
        val postscripts: List<Postscript>?
) : Page(), Thankable, Ignorable, Favable {
    @IgnoredOnParcel
    val hasInfo = member != null

    override fun getTitle(): String? {
        return title
    }

    override fun getUrl(): String {
        return buildUrlFromId(id)
    }

    override fun getIgnoreUrl(): String {
        return String.format("%s/ignore/topic/%d", RequestHelper.BASE_URL, id)
    }

    override fun getThankUrl(): String {
        return String.format("%s/thank/topic/%d", RequestHelper.BASE_URL, id)
    }

    override fun getFavUrl(token: String): String {
        return String.format("%s/favorite/topic/%d?t=%s", RequestHelper.BASE_URL, id, token)
    }

    override fun getUnFavUrl(token: String): String {
        return String.format("%s/unfavorite/topic/%d?t=%s", RequestHelper.BASE_URL, id, token)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        if (!super.equals(other)) return false

        other as Topic

        if (id != other.id) return false
        if (title != other.title) return false
        if (content != other.content) return false
        if (replyCount != other.replyCount) return false
        if (member != other.member) return false
        if (node != other.node) return false
        if (replyTime != other.replyTime) return false
        if (isFavored != other.isFavored) return false
        if (postscripts != other.postscripts) return false
        if (hasInfo != other.hasInfo) return false

        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + id
        result = 31 * result + (title?.hashCode() ?: 0)
        result = 31 * result + (content?.hashCode() ?: 0)
        result = 31 * result + replyCount
        result = 31 * result + (member?.hashCode() ?: 0)
        result = 31 * result + (node?.hashCode() ?: 0)
        result = 31 * result + (replyTime?.hashCode() ?: 0)
        result = 31 * result + isFavored.hashCode()
        result = 31 * result + (postscripts?.hashCode() ?: 0)
        result = 31 * result + hasInfo.hashCode()
        return result
    }

    override fun toString(): String {
        return "Topic(id=$id, title=$title)"
    }

    class Builder {
        var id: Int by Delegates.notNull()
        var title: String? = null
        var content: List<ContentBlock>? = null
        var replyCount: Int = 0
        var member: Member? = null
        var node: Node? = null
        var replyTime: String? = null
        var isFavored: Boolean = false
        var postscripts: List<Postscript>? = null

        val hasInfo: Boolean
            get() = member != null

        fun build(): Topic {
            return Topic(
                    id,
                    title,
                    content,
                    replyCount,
                    member,
                    node,
                    replyTime,
                    isFavored,
                    postscripts
            )
        }
    }

    fun toBuilder(): Builder {
        return Builder().also {
            it.id = this.id
            it.title = this.title
            it.content = this.content
            it.replyCount = this.replyCount
            it.member = this.member
            it.node = this.node
            it.replyTime = this.replyTime
            it.isFavored = this.isFavored
            it.postscripts = this.postscripts
        }
    }

    companion object {
        private val PATTERN = Pattern.compile("/t/(\\d+?)(?:\\W|$)")

        fun buildUrlFromId(id: Int): String {
            return RequestHelper.BASE_URL + "/t/" + id.toString()
        }

        @JvmStatic
        fun getIdFromUrl(url: String): Int {
            val matcher = PATTERN.matcher(url)
            if (!matcher.find()) {
                throw IllegalArgumentException("match id for topic failed: $url")
            }
            val idStr = matcher.group(1)
            return Integer.parseInt(idStr)
        }
    }
}
